skmultiflow/classification/trees/hoeffding_adaptive_tree.py

```diff
@@ -0,0 +1,489 @@
+
+            dist_sum = sum(dist.values())  # sum all values in dictionary
+
+            if dist_sum * self.get_error_estimation() * self.get_error_estimation() > 0.0:
+                normalize(dist_sum * self.get_error_estimation() * self.get_error_estimation(), dist)
+
+            return dist
+
+        # Override NewNode, New for option votes
+        def filter_instance_to_leaves(self, X, split_parent, parent_branch, update_splitter_counts, found_nodes=None):
+            if found_nodes is None:
+                found_nodes = []
+            found_nodes.append(HoeffdingTree.FoundNode(self, split_parent, parent_branch))
+
+    # =============================================
+    # == Hoeffding Adaptive Tree implementation ===
+    # =============================================
+
+    def __init__(self, *args, **kwargs):
+
+        super(HAT, self).__init__(*args, **kwargs)
+        self._alternate_trees = 0
+        self._pruned_alternate_trees = 0
+        self._switch_alternate_trees = 0
+        self._tree_root = None  # CHECK I not utilize _tree_root of HoeffdingTree. OK
+
+    def reset(self):
+        self._alternate_trees = 0
+        self._pruned_alternate_trees = 0
+        self._switch_alternate_trees = 0
+        self._tree_root = None
+
+    # Override HoeffdingTree/BaseClassifier
+    def partial_fit(self, X, y, classes=None, weight=None):
+        """ partial_fit
+
+        Trains the model on samples X and targets y.
+
+        Parameters
+        ----------
+        X: Numpy.ndarray of shape (n_samples, n_features)
+            Data instances.
+
+        y: Array-like
+            Contains the classification targets for all samples in X.
+
+        classes: Not used.
+
+        weight: Float or Array-like
+            Instance weight. If not provided, uniform weights are assumed.
+
+        Returns
+        -------
+        self
+
+        """
+        if y is not None:
+            if weight is None:
+                weight = np.array([1.0])
+            row_cnt, _ = get_dimensions(X)
+            wrow_cnt, _ = get_dimensions(weight)
+            if row_cnt != wrow_cnt:
+                weight = [weight[0]] * row_cnt
+            for i in range(row_cnt):
+                if weight[i] != 0.0:
+                    self._partial_fit(X[i], y[i], weight[i])
+
+    # Override HoeffdingTree
+    def _partial_fit(self, X, y, weight):
+
+        if self._tree_root is None:
+            self._tree_root = self._new_learning_node()
+            self._active_leaf_node_cnt = 1
+        self._tree_root.learn_from_instance(X, y, weight, self, None, -1)
+
+    def filter_instance_to_leaves(self, X, split_parent, parent_branch, update_splitter_counts):
+        nodes = []
+        self._tree_root.filter_instance_to_leaves(X, split_parent, parent_branch, update_splitter_counts, nodes)
+        return nodes
+
+    # Override HoeffdingTree
```

Comment:
Minor change suggestion, initialize result at the beginning and move the return out of the *if* block (no need for *else*):
``` python
def get_votes_for_instance(self, X):
    result = {}
    if self._tree_root is not None:
        found_node = self.filter_instance_to_leaves(X, None, -1, False)
        for fn in found_node:
            if fn.parent_branch != -999:
                leaf_node = fn.node
                if leaf_node is None:
                    leaf_node = fn.parent
                dist = leaf_node.get_class_votes(X, self)
                result.update(dist)  # add elements to dictionary
    return result
```

- [ ] Remove unused variable ```predictionPaths```